SwiftKitCore/src/main/java/org/swift/swiftkit/core/SwiftLegacyFuture.java

```diff
@@ -0,0 +1,149 @@
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of Swift.org project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+package org.swift.swiftkit.core;
+
+import java.util.Deque;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+
+/**
+ * A simple completable Future for platforms that do not support CompletableFuture.
+ *
+ * @param <T> The result type
+ */
+public class SwiftLegacyFuture<T> implements Future<T> {
+    private final Object lock = new Object();
+    private final Deque<Runnable> callbacks = new ConcurrentLinkedDeque<>();
+    private volatile T result;
+    private volatile Throwable thrownException;
+    private final AtomicBoolean isCompleted = new AtomicBoolean(false);
+
+    private void addCallback(Runnable action) {
+        callbacks.add(action);
+        if (isDone()) {
+            runCallbacks();
+        }
+    }
+
+    public <U> Future<U> thenApply(Function<? super T, ? extends U> fn) {
+        SwiftLegacyFuture<U> newFuture = new SwiftLegacyFuture<>();
+        Runnable callback = () -> {
+            if (this.thrownException != null) {
+                newFuture.completeExceptionally(this.thrownException);
+            } else {
+                try {
+                    U newResult = fn.apply(this.result);
+                    newFuture.complete(newResult);
+                } catch (Throwable t) {
+                    newFuture.completeExceptionally(t);
+                }
+            }
+        };
+        addCallback(callback);
+        return newFuture;
+    }
+
+    public boolean complete(T value) {
+        if (isCompleted.compareAndSet(false, true)) {
+            this.result = value;
+            synchronized (lock) {
+                lock.notifyAll();
+            }
+            runCallbacks();
+            return true;
+        }
+        return false;
+    }
+
+    public boolean completeExceptionally(Throwable ex) {
+        if (isCompleted.compareAndSet(false, true)) {
+            this.thrownException = ex;
+            synchronized (lock) {
+                lock.notifyAll();
+            }
+            runCallbacks();
+            return true;
+        }
+        return false;
+    }
+
+    private void runCallbacks() {
+        Runnable callback;
+        while ((callback = callbacks.pollFirst()) != null) {
+            callback.run();
+        }
+    }
+
+    @Override
+    public boolean cancel(boolean mayInterruptIfRunning) {
+        return false;
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return false;
+    }
+
+    @Override
+    public boolean isDone() {
+        return isCompleted.get();
+    }
+
+    @Override
+    public T get() throws InterruptedException, ExecutionException {
```

Comment:
The get may be called frequently so it's worth adding a fast path when we are completed so we don't take the lock then